ocaml-lsp-server/src/diagnostics.ml

```diff
@@ -0,0 +1,100 @@
+open Import
+
+module Uri = struct
+  include Uri
+
+  let compare x y = Ordering.of_int (Uri.compare x y)
+end
+
+module Uri_c = Comparable.Make (Uri)
+module Uri_set = Uri_c.Set
+
+type dune_status =
+  | Inactive
+  | Connected
+  | Disconnected
+
+type t =
+  { dune : (Drpc.Diagnostic.Id.t, Uri.t * Diagnostic.t) Table.t
+  ; mutable dune_status : dune_status
+  ; workspace_root : Uri.t Lazy.t
+  ; merlin : (Uri.t, Diagnostic.t list) Table.t
+  ; send : PublishDiagnosticsParams.t list -> unit Fiber.t
+  ; mutable dirty_uris : Uri_set.t
+  }
+
+let workspace_root t = Lazy.force t.workspace_root
+
+module Id = struct
+  include Drpc.Diagnostic.Id
+
+  let equal x y = compare x y = 0
+
+  let to_dyn = Dyn.Encoder.opaque
+end
+
+let create send ~workspace_root =
+  { dune = Table.create (module Id) 32
+  ; merlin = Table.create (module Uri) 32
+  ; dirty_uris = Uri_set.empty
+  ; dune_status = Inactive
+  ; send
+  ; workspace_root
+  }
+
+let update_dune_status t status =
+  t.dirty_uris <- Uri_set.add t.dirty_uris (Lazy.force t.workspace_root);
+  t.dune_status <- status
+
+let dune_status_diagnostic t =
+  match t.dune_status with
+  | Disconnected ->
+    Some
+      (Diagnostic.create ~severity:Information ~source:"dune"
+         ~range:Range.first_line
+         ~message:
+           "Dune diagnostic status may be stale. Please run dune in watch mode\n\
+            to see up to date diagnostics"
+         ())
+  | _ -> None
+
+let send t =
+  let pending = Table.create (module Uri) 32 in
+  dune_status_diagnostic t
+  |> Option.iter ~f:(fun d ->
+         Table.add_exn pending (Lazy.force t.workspace_root) [ d ]);
+  Uri_set.iter t.dirty_uris ~f:(fun uri ->
+      let diagnostics = Table.Multi.find t.merlin uri in
+      Table.set pending uri diagnostics);
+  Table.iter t.dune ~f:(fun (uri, diagnostic) ->
+      if Uri_set.mem t.dirty_uris uri then
+        Table.Multi.cons pending uri diagnostic);
+  t.dirty_uris <- Uri_set.empty;
+  Table.foldi pending ~init:[] ~f:(fun uri diagnostics acc ->
```

Comment:
It's strange that `foldi` does not pass the index to `~f`

Reply:
The index in `Table.foldi` is the key in the hash table. In this case that's passed as `uri`.